node/lib/util/sequencer_state_util.js

```diff
@@ -0,0 +1,237 @@
+const CURRENT_COMMIT_FILE = "CURRENT_COMMIT";
+
+/**
+ * Return the contents of the file in the sequencer directory from the
+ * specified `gitDir` having the specified `name`, or null if the file cannot
+ * be read.
+ *
+ * @param {String} gitDir
+ * @param {String} name
+ * @return {String|null}
+ */
+exports.readFile = co.wrap(function *(gitDir, name) {
+    assert.isString(gitDir);
+    assert.isString(name);
+    const filePath = path.join(gitDir, SEQUENCER_DIR, name);
+    try {
+        return yield fs.readFile(filePath, "utf8");
+    }
+    catch (e) {
+        return null;
+    }
+});
+
+/**
+ * Read the `CommitAndRef` object from the specified `fileName` in the
+ * sequencer director in the specified `gitDir` if it exists, or null if it
+ * does not.
+ *
+ * @param {String} gitDir
+ * @param {String} fileName
+ * @return {CommitAndRef|null}
+ */
+exports.readCommitAndRef = co.wrap(function *(gitDir, fileName) {
+    assert.isString(gitDir);
+    assert.isString(fileName);
+    const content = yield exports.readFile(gitDir, fileName);
+    if (null !== content) {
+        const lines = content.split("\n");
+
+        // always an extra entry after the split
+
+        const numLines = lines.length - 1; 
+        if (1 === numLines || 2 === numLines) {
+            const ref = 2 === numLines ? lines[1] : null;
+            return new CommitAndRef(lines[0], ref);
+        }
+    }
+    return null;
+});
+
+/**
+ * Return the array of commit stored in the specified `gitDir`, or null if
+ * the file is missing or malformed.
+ *
+ * @param {String} gitDir
+ * @return {[String]|null}
+ */
+exports.readCommits = co.wrap(function *(gitDir) {
+    assert.isString(gitDir);
+    const content = yield exports.readFile(gitDir, COMMITS_FILE);
+    if (null !== content) {
+        const lines = content.split("\n");
+        const nonEmpty = lines.filter(line => line.length > 0);
+        if (0 !== nonEmpty.length) {
+            return nonEmpty;
+        }
+    }
+    return null;
+});
+
+/**
+ * Return the index of the current commit stored in the specified `gitDir`
+ * having the specified `numCommits` number of commits  or null if the file is
+ * missing, malformed, our out-of-bounds.
+ *
+ * @param {String} gitDir
+ * @param {Number} numCommits
+ * @return {Number}
+ */
+exports.readCurrentCommit = co.wrap(function *(gitDir, numCommits) {
+    assert.isString(gitDir);
+    assert.isNumber(numCommits);
+    const content = yield exports.readFile(gitDir, CURRENT_COMMIT_FILE);
+    if (null !== content) {
+        const lines = content.split("\n");
+        if (0 !== lines.length) {
+            const index = Number.parseInt(lines[0]);
+            if (!Number.isNaN(index) && index < numCommits) {
+                return index;
+            }
+        }
+    }
+    return null;
+});
+
+/**
+ * Return the sequencer state if it exists in the specified `gitDir`, or null
+ * if it is missing or malformed.
+ * TODO: emit diagnostic when malformed?
```

Comment:
raise exception if malformed -- let the caller say, "hey, your sequencer state is malformed.  This shouldn't happen.  You can fix it by deleting .git/meta-sequencer and then using the reflog to get back to where you started."

(ok to leave like this for this pr)

Reply:
Yes, I want to think about this some more.